### PR TITLE
New expression: count_leading_zeros_exprt

### PR DESCRIPTION
**Pull request by tautschnig**
Rather than ad-hoc handling __builtin_clz and __lzcnt (and their
variants) in the C front-end, make counting-leading-zeros available
across the code base.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/ansi-c/c_typecheck_expr.cpp [hunk 1]
```diff
@@ -2689,27 +2689,22 @@ exprt c_typecheck_baset::do_special_functions(
...
 
     if(
-      !int_constant.has_value() || *int_constant == 0 ||
-      argument.type().id() != ID_unsignedbv)
+      config.ansi_c.mode == configt::ansi_ct::flavourt::GCC ||
+      config.ansi_c.mode == configt::ansi_ct::flavourt::CLANG)
     {
-      return nil_exprt{};
+      simplify(clz, *this);
     }
```
Comment (kroening) [inline src/ansi-c/c_typecheck_expr.cpp]: I wouldn't simplify these up-front -- the simplifier is called by the front-end whenever a constant is needed.

File: src/goto-programs/goto_clean_expr.cpp [hunk 2]
```diff
@@ -429,6 +438,22 @@ void goto_convertt::clean_expr(
...
     expr = to_unary_expr(expr).op();
   }
+  else if(expr.id() == ID_count_leading_zeros)
+  {
+    count_leading_zeros_exprt &clz = to_count_leading_zeros_expr(expr);
+    if(!clz.zero_permitted())
+    {
+      clz.zero_permitted(true);
+      const source_locationt source_location = clz.source_location();
+
```
Comment (kroening) [inline src/goto-programs/goto_clean_expr.cpp]: I would go stronger than the nondet: The wording in the gcc docs is that __builtin_clz is undefined when the argument is zero. Assuming they use the meaning of "undefined" as in the ISO C standard, then this is worth an assertion, to be generated in goto_check.

File: src/ansi-c/c_typecheck_expr.cpp (hunk 1 above)
Comment (tautschnig) [reply src/ansi-c/c_typecheck_expr.cpp]: Ah, true, will fix.

File: src/goto-programs/goto_clean_expr.cpp (hunk 2 above)
Comment (tautschnig) [reply src/goto-programs/goto_clean_expr.cpp]: Hmm, true, then I could perhaps demote the `ID_zero_permitted` to `ID_C_zero_permitted`?

Comment (kroening) [reply src/goto-programs/goto_clean_expr.cpp]: I don't think you need it at all. We don't have anything equivalent for any of the other expressions that are only partially defined.

Comment (tautschnig) [reply src/goto-programs/goto_clean_expr.cpp]: But we need to distinguish whether we're handling MSVC's `__lzcnt` or `__builtin_clz`. The former is defined for all values, the latter isn't defined for `0`.

Comment (kroening) [reply src/goto-programs/goto_clean_expr.cpp]: Ah, right. Then I'd make it `ID_zero_permitted`; a comment would be too weak, given that the semantics do differ.

Comment (tautschnig) [reply src/goto-programs/goto_clean_expr.cpp]: Except they only sort-of do: it seems that in practice it's almost always ok, it's just that GCC doesn't provide a guarantee: https://stackoverflow.com/questions/19527897/how-undefined-are-builtin-ctz0-or-builtin-clz0. How about the solution that I just pushed?

**Review by martin-cs — APPROVED** [review]
Seems reasonable if we want to go that way.

**Review by kroening — APPROVED** [review]

**Pull request merged**